main/provider/index.ts

```diff
@@ -762,38 +589,30 @@ export class Provider extends EventEmitter {
       const params = payload.params
       if (!params || !params[0]) throw new Error('Params not supplied')
       
-      const type = 'ethereum'
 
       const chainId = parseInt(params[0].chainId)
       if (!Number.isInteger(chainId)) throw new Error('Invalid chain id')
 
       // Check if chain exists 
-      const exists = Boolean(store('main.networks', type, chainId))
+      const exists = Boolean(store('main.networks', 'ethereum', chainId))
       if (exists === false) throw new Error('Chain does not exist')
 
-      if (store('main.currentNetwork.id') === chainId) return res({ id: payload.id, jsonrpc: '2.0', result: null })
-
-      const handlerId = this.addRequestHandler(res)
+      const originId = uuidv5(payload._origin, uuidv5.DNS)
+      const { chainId: currentChain, type } = store('main.origins', originId)
       
-      // Ask user if they want to switch chains
-      accounts.addRequest({
-        handlerId,
-        type: 'switchChain',
-        chain: { 
-          type, 
-          id: params[0].chainId
-        },
-        account: (accounts.getAccounts() || [])[0],
-        origin: payload._origin,
-        payload
-      } as SwitchChainRequest, res)
+      if (currentChain !== chainId) {
+        store.switchOriginChain(originId, chainId, type)
+        this.chainChanged(chainId, payload._origin)
```

Comment:
will this cause 2 events to be emitted, one here and one via the observer?

Reply:
good catch, just fixed